File: internal/engine/criar.go
package engine

import (
	"fmt"

	"monitor-versoes/internal/domain"
	"monitor-versoes/internal/services"
)

// Criar monta uma versao do zero (§5). Branch nova e nao publicada - rebuild
// idempotente e permitido ate a primeira publicacao (§6), mas esta operacao
// so cria; recriar do zero e responsabilidade do chamador (remover a worktree
// antes de chamar Criar de novo).
func Criar(d Deps, versao string) (IncrementResult, error) {
	gate := services.PublicationGate{Git: d.Git}
	publicada, err := gate.Publicada(versao)
	if err != nil {
		return IncrementResult{}, err
	}
	if publicada {
		return IncrementResult{}, fmt.Errorf("versao %s ja publicada - use Incrementar", versao)
	}

	baseResolver := services.BaseResolver{Git: d.Git}
	base, err := baseResolver.Resolve(versao)
	if err != nil {
		return IncrementResult{}, err
	}

	if err := d.Git.WorktreeAdd(versao, base.Ref); err != nil {
		return IncrementResult{}, err
	}

	tipo, err := domain.InferirTipo(versao)
	if err != nil {
		return IncrementResult{}, err
	}
	lockStore := services.LockStore{Git: d.Git}
	lockInicial := domain.Lock{Versao: versao, Tipo: tipo, Base: base, Tasks: domain.TargetSet{}}
	if err := lockStore.Escrever(versao, lockInicial); err != nil {
		return IncrementResult{}, err
	}

	return Incrementar(d, versao)
}

File: motor/engine/criar.py
"""Composicao de uma versao nova: branch, VERSAO e registro no estado."""

from __future__ import annotations

from motor.domain.types import VersaoInfo
from motor.domain.version import inferir_tipo
from motor.engine.atualizar import AtualizarResult, atualizar
from motor.engine.deps import Deps
from motor.errors import RecusaDeInvariante
from motor.services.base_resolver import BaseResolver
from motor.services.publication_gate import PublicationGate


def criar(deps: Deps, versao: str) -> AtualizarResult:
    """Monta uma versao do zero (spec §5). Branch nova e nao publicada.

    So cria: recriar do zero e responsabilidade do chamador, que precisa remover
    a worktree e a branch antes de chamar de novo — `worktree_add` recusa branch
    existente, entao um retry apos lote BLOCKED cai aqui, nao no atualizar.
    """
    # A trava e a base saem do ref store local (`git tag -l`,
    # `list_version_branches`): sem buscar primeiro, uma versao liberada em
    # outra maquina passaria pela trava e ganharia branch nova aqui.
    deps.git.fetch("origin")

    gate = PublicationGate(git=deps.git)
    if gate.publicada(versao):
        raise RecusaDeInvariante(f"versao {versao} ja publicada - use atualizar")

    base = BaseResolver(git=deps.git).resolve(versao)

    deps.git.worktree_add(versao, base.commit)
    deps.git.write_file(
        versao, "VERSAO", f"{versao}\n".encode(), f"Atualiza VERSAO para {versao}"
    )

    # A versao entra no estado aqui; o verificar dentro do atualizar preenche
    # as atribuicoes.
    deps.estado.registrar_versao(
        deps.repo,
        VersaoInfo(
            numero=versao,
            tipo=inferir_tipo(versao),
            base_ref=base.ref,
            base_commit=base.commit,
        ),
    )

    return atualizar(deps, versao)
